synapse/storage/databases/main/appservice.py

```diff
@@ -72,6 +74,21 @@ def __init__(
         )
         self.exclusive_user_regex = _make_exclusive_regex(self.services_cache)
 
+        def get_max_as_txn_id(txn: Cursor) -> int:
+            txn.execute(
+                "SELECT COALESCE(max(txn_id), 0) FROM application_services_txns"
+            )
```

Comment:
problem here is that it will mean a table-scan of `application_services_txns` on startup - which could be very slow.

That said, this is only used on SQLite, and tbh anyone running a busy AS with sqlite probably needs to move over to postgres anyway.

So perhaps we can just log a warning ("falling back to slow query, you need to port to postgres") if we ever hit this path.

Reply:
https://github.com/matrix-org/synapse/pull/12209/commits/cceb2b9caa4bfd0137e46b3bf4328d151084a6cb